[#682] Don't lock the file when open for reading.

adiroiban opened the pull request:
Scope
=====

Fixes #682

This tries to get the same behaviour over linux and Windows when opening a file for reading.

The default Python code locks the file on Windows, but the file is not locked on Linux.

Changes
=======

Implement custom Windows open code.


How to try and test the changes
===============================

reviewers: @danuker 

Check that changes make sense

I will create a separate PR for chevah/server in which this will be used and we can do more testing.

adiroiban commented:
This is ready for a review.

No need to run manual tests.

We have the automated tests that are also executed on Windows

needs-review

danuker commented on chevah/compat/tests/normal/test_filesystem.py:
```diff
@@ -1715,6 +1715,95 @@ def test_openFileForReading_no_write(self):
+                mode=0,
+                )
+
+            self.filesystem.deleteFile(segments)
```
Maybe also check that the file WAS deleted (such as with `os.stat`).

adiroiban replied:
I think this was ok.
If `filesystem.deleteFile` is not doing what it should, we should have a separate test and fix for the `delteFile` method.

To be extra safe, I went with:

```diff
            self.filesystem.deleteFile(segments)
+           self.assertFalse(self.filesystem.exists(segments))
```

danuker commented on chevah/compat/tests/normal/test_filesystem.py:
```diff
@@ -1715,6 +1715,95 @@ def test_openFileForReading_no_write(self):
+    @conditionals.onOSFamily('nt')
+    def test_fdRead_os_error(self):
+        """
+        This is a low-level Windows tests.
```
```suggestion
        This is a low-level Windows test.
```

danuker commented on chevah/compat/tests/normal/test_filesystem.py:
```diff
@@ -1715,6 +1715,95 @@ def test_openFileForReading_no_write(self):
+        Under normal update, this function should never fail as we have some
+        higher API guards.
+
+        This tests is here to make sure that if it fails, it raises a generic
```
```suggestion
        This test is here to make sure that if it fails, it raises a generic
```

danuker commented on chevah/compat/tests/normal/test_filesystem.py:
```diff
@@ -1715,6 +1715,95 @@ def test_openFileForReading_no_write(self):
+                os.close(fd)
+
+        # Just check the result after close.
+        self.assertEqual(content, mk.fs.getFileContent(segments))
```
I'd like an `OPEN_READ_WRITE` test also, as I said in `chevah/compat/nt_filesystem.py`.

adiroiban replied:
Added.  Thanks! Good catch!

danuker reviewed (changes requested):
I'm a bit worried about the `OPEN_READ_WRITE` case.

adiroiban commented on chevah/compat/nt_filesystem.py:
```diff
@@ -978,3 +979,64 @@ def hasGroup(self, segments, group):
+
+        self._requireFile(segments)
+        with self._convertToOSError(path), self._impersonateUser():
+            if (
```
Yes. On Windows, os.open() from STDLib does an exclusive lock and there is no flag to disable this behaviour :(

adiroiban commented on chevah/compat/nt_filesystem.py:
```diff
@@ -978,3 +979,64 @@ def hasGroup(self, segments, group):
+        self._requireFile(segments)
+        with self._convertToOSError(path), self._impersonateUser():
+            if (
+                flags & self.OPEN_READ_ONLY == self.OPEN_READ_ONLY
+                and flags & self.OPEN_WRITE_ONLY != self.OPEN_WRITE_ONLY
```
Yes. Good catch. I patched it as best as I could :(

adiroiban commented:
Many thanks for another great review.

I don't know why coverage is not happy, but we can deal with it later.

I hope I have address all the major issues

needs-review

danuker reviewed (approved):
Looks good! Thanks.
I pushed a typo fix.

adiroiban commented on chevah/compat/nt_filesystem.py:
```diff
@@ -978,3 +999,69 @@ def hasGroup(self, segments, group):
+        desired_access = win32file.GENERIC_READ
+        share_mode = (
+            win32file.FILE_SHARE_DELETE |
+            win32file.FILE_SHARE_WRITE |
```
@danuker in previous version I was missing FILE_SHARE_WRITE ... so it was failing to open a file that was already opened for writing.

This should get a behaviour closer to what python os.open does.

I tried to do a quick search for python implementation of `open` but I could not find it

adiroiban commented:
Thanks for the review.

There were a few changes and updated tests.

The failing tests were observed on chevah/server

I have implemented more tests into chevah/compat. I hope this should give us a better coverage for the intended behaviour

needs-review

danuker reviewed (approved):
Looks good. I pushed a few typo fixes.

adiroiban commented:
Thanks. I have merged this without a new-release.

The 0.65.5 is already used in chevah/server for a long time
